test/cmocka/src/audio/eq_iir/eq_iir_process.c

```diff
@@ -0,0 +1,450 @@
+	ss = &sb->stream;
+	frames = MIN(audio_stream_get_free_frames(ss), frames_max);
+	samples = frames * ss->channels;
+	for (i = 0; i < samples; i++) {
+		x = audio_stream_write_frag_s16(ss, i);
+		*x = sat_int16(Q_SHIFT_RND(chirp_2ch[buffer_fill_data.idx++], 31, 15));
+		samples_processed++;
+		if (buffer_fill_data.idx == CHIRP_2CH_LENGTH) {
+			td->continue_loop = false;
+			break;
+		}
+	}
+
+	if (samples_processed > 0) {
+		bytes_total = samples_processed * audio_stream_sample_bytes(ss);
+		comp_update_buffer_produce(sb, bytes_total);
+	}
+}
+
+static void verify_sink_s16(struct test_data *td)
+{
+	struct comp_dev *dev = td->dev;
+	struct comp_buffer *sb;
+	struct audio_stream *ss;
+	int32_t delta;
+	int32_t ref;
+	int32_t out;
+	int16_t *x;
+	int samples;
+	int frames;
+	int i;
+
+	sb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	ss = &sb->stream;
+	frames = audio_stream_get_avail_frames(ss);
+	samples = frames * ss->channels;
+	for (i = 0; i < samples; i++) {
+		x = audio_stream_read_frag_s16(ss, i);
+		out = *x;
+		ref = sat_int16(Q_SHIFT_RND(chirp_iir_ref_2ch[buffer_verify_data.idx++], 31, 15));
+		delta = ref - out;
+		if (delta > ERROR_TOLERANCE_S16 || delta < -ERROR_TOLERANCE_S16)
+			assert_int_equal(out, ref);
+	}
+
+	if (frames > 0)
+		comp_update_buffer_consume(sb, frames * audio_stream_frame_bytes(ss));
+}
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+static void fill_source_s24(struct test_data *td, int frames_max)
+{
+	struct comp_dev *dev = td->dev;
+	struct comp_buffer *sb;
+	struct audio_stream *ss;
+	int32_t *x;
+	int bytes_total;
+	int samples;
+	int frames;
+	int i;
+	int samples_processed = 0;
+
+	sb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	ss = &sb->stream;
+	frames = MIN(audio_stream_get_free_frames(ss), frames_max);
+	samples = frames * ss->channels;
+	for (i = 0; i < samples; i++) {
+		x = audio_stream_write_frag_s32(ss, i);
+		*x = sat_int24(Q_SHIFT_RND(chirp_2ch[buffer_fill_data.idx++], 31, 23));
+		samples_processed++;
+		if (buffer_fill_data.idx == CHIRP_2CH_LENGTH) {
+			td->continue_loop = false;
+			break;
+		}
+	}
+
+	if (samples_processed > 0) {
+		bytes_total = samples_processed * audio_stream_sample_bytes(ss);
+		comp_update_buffer_produce(sb, bytes_total);
+	}
+}
+
+static void verify_sink_s24(struct test_data *td)
+{
+	struct comp_dev *dev = td->dev;
+	struct comp_buffer *sb;
+	struct audio_stream *ss;
+	int32_t delta;
+	int32_t ref;
+	int32_t out;
+	int32_t *x;
+	int samples;
+	int frames;
+	int i;
+
+	sb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	ss = &sb->stream;
+	frames = audio_stream_get_avail_frames(ss);
+	samples = frames * ss->channels;
+	for (i = 0; i < samples; i++) {
+		x = audio_stream_read_frag_s32(ss, i);
+		out = *x;
+		ref = sat_int24(Q_SHIFT_RND(chirp_iir_ref_2ch[buffer_verify_data.idx++], 31, 23));
+		delta = ref - out;
```

Comment:
Shortcoming here, need to make sure there's no 24 bit overflow in int32_t out. Shifting to MSB and back will fix this.

Reply:
This is now fixed with the left right shifts. I just checked that it detected the mistake I had in the IIR patch.